### PR TITLE
Add default config

### PR DESCRIPTION
**Pull request by gferraro**

### CODE REVIEW

**Comment by mjs** [comment]
Conflicts here

File: .travis.yml [hunk 1]
```diff
@@ -18,6 +18,8 @@ install:
   - cd ..
   - docker cp db-test-seed.sql cacophony-api:/db-seed.sql
 before_script:
+  - sudo mkdir /etc/cacophony
+  - sudo chmod 777 /etc/cacophony
```
Comment (mjs) [inline .travis.yml]: Does this mean that tests are looking in /etc/cacophony? That's not going to be reliable, especially on developer machines. See below for more on this.

File: config.go [hunk 2]
```diff
@@ -29,42 +29,62 @@ import (
 	yaml "gopkg.in/yaml.v2"
 )
 
+const deviceConfigPath = "/etc/cacophony/device.yml"
```
Comment (mjs) [inline config.go]: Can you please use the ".yaml" extension. That's what our other config files use. Both .yml and .yaml are correct for YAML files but let's keep things consistent.

File: api_test.go [hunk 3]
```diff
@@ -319,6 +331,11 @@ func runMultipleRegistrations(configFile string, count int) (int, chan string) {
 	return count, messages
 }
 
+func clearConfig() {
+	_ = os.Remove(deviceConfigPath)
+	_ = os.Remove(privConfigFilename(deviceConfigPath))
```
Comment (mjs) [inline api_test.go]: As mentioned above, having tests write out to /etc on people's machines like this isn't great.

A good way to handle this is to use the [afero package](https://github.com/spf13/afero). It provides a filesystem abstraction which can be swapped out for a fake memory based "filesystem" during tests. This makes tests much cleaner too.

An example of a simple package that uses afero like this is here: https://github.com/jumptrading/influx-spout/tree/master/config

Some key lines in that code:
* https://github.com/jumptrading/influx-spout/blob/master/config/config.go#L305
* https://github.com/jumptrading/influx-spout/blob/master/config/config.go#L278
* https://github.com/jumptrading/influx-spout/blob/master/config/config_small_test.go#L325

Use afero will allow things like `createTestConfig` to be simpler too.

File: config.go [hunk 4]
```diff
@@ -29,42 +29,62 @@ import (
...
+}
 
-	return ioutil.WriteFile(filename, buf, 0600)
+// LoadConfig from deviceConfigPath
+func LoadConfig() (*Config, error) {
+	return ParseConfigFile(deviceConfigPath)
 }
 
-//ParseConfig takes supplied filename and returns a parsed Config struct
+//ParseConfig takes supplied filenam)e and returns a parsed Config struct
```
Comment (mjs) [inline config.go]: filename

File: config.go [hunk 5]
```diff
@@ -29,42 +29,62 @@ import (
...
+
+	if (conf.Group == "" || conf.DeviceName == "") && conf.DeviceID == 0 {
+		return errors.New("device-info missing")
 	}
 	return nil
 }
 
-//SaveToFile writes config to supplied filename
-func (conf *Config) SaveToFile(filename string) error {
+//Save writes config to deviceConfig
```
Comment (mjs) [inline config.go]: replace `deviceConfig` with `device.yaml` here?

File: api.go [hunk 6]
```diff
@@ -86,14 +86,32 @@ func (api *CacophonyAPI) JustRegistered() bool {
 	return api.justRegistered
 }
 
-// NewAPIFromConfig prases the supplied configFile and creates a new CacophonyAPI with the configFile information
+// NewAPIFromConfig parses file name into deviceConfigPath if it doesn't exist or overwrite set
+// and creates a new CacophonyAPI with the configFile information
 // and saves the generated password to privConfigFileName(configFile)
-func NewAPIFromConfig(configFile string) (*CacophonyAPI, error) {
-	conf, err := ParseConfigFile(configFile)
+func NewAPIFromConfig(filename string, overwrite bool) (*CacophonyAPI, error) {
```
Comment (mjs) [inline api.go]: How would the caller decide what to pass for the `overwrite` value? Would it better to always update device.yaml?

File: config.go [hunk 7]
```diff
@@ -29,42 +29,62 @@ import (
...
+func (conf *Config) Save() error {
 	buf, err := yaml.Marshal(&conf)
 	if err != nil {
 		return err
 	}
+	return ioutil.WriteFile(deviceConfigPath, buf, 0600)
```
Comment (mjs) [inline config.go]: There's a small chance that multiple API client might try to write to device.yaml at the same time. There should some file locking as there is for password file.

File: api_test.go [hunk 8]
```diff
@@ -331,20 +333,14 @@ func runMultipleRegistrations(configFile string, count int) (int, chan string) {
...
-	_ = os.Remove(deviceConfigPath)
-	_ = os.Remove(privConfigFilename(deviceConfigPath))
-}
-
-func removeTestConfig(configFile string) {
-	_ = os.Remove(configFile)
-	_ = os.Remove(privConfigFilename(configFile))
+func removeConfig() {
+	Fs.Remove(deviceConfigPath)
+	Fs.Remove(registeredConfigPath)
```
Comment (mjs) [inline api_test.go]: There's no need for this if each test resets `Fs`. Remember that the filesystem is being faked in memory during tests.

Comment (gferraro) [reply api_test.go]: True

**Pull request merged**